File: Source/UnrealBowlingUSharp/Private/MyBowlPlayerComponent.cpp
// Fill out your copyright notice in the Description page of Project Settings.


#include "Public/MyBowlPlayerComponent.h"
#include "Kismet/GameplayStatics.h"
#include "BowlingBallComponent.h"
#include "BowlGameMasterComponent.h"
#include "BowlGameModeComponent.h"
#include "GameFramework/GameModeBase.h"

// Sets default values for this component's properties
UMyBowlPlayerComponent::UMyBowlPlayerComponent()
{
	// Set this component to be initialized when the game starts, and to be ticked every frame.  You can turn these features
	// off to improve performance if you don't need them.
	PrimaryComponentTick.bCanEverTick = true;
	BallFollowLimitDistance = 3200.0f;
	// ...
}

#pragma region Component Getters
UBowlGameMasterComponent* UMyBowlPlayerComponent::GetGameMaster()
{
	if (bowlGameMaster == nullptr)
	{
		auto _gamemode = UGameplayStatics::GetGameMode(this);
		if (_gamemode != nullptr)
		{
			bowlGameMaster = Cast<UBowlGameMasterComponent>(_gamemode->GetComponentByClass(UBowlGameMasterComponent::StaticClass()));
		}
	}
	return bowlGameMaster;
}

UBowlGameModeComponent* UMyBowlPlayerComponent::GetBowlGameMode()
{
	if (bowlGameMode == nullptr)
	{
		auto _gamemode = UGameplayStatics::GetGameMode(this);
		if (_gamemode != nullptr)
		{
			bowlGameMode = Cast<UBowlGameModeComponent>(_gamemode->GetComponentByClass(UBowlGameModeComponent::StaticClass()));
		}
	}
	return bowlGameMode;
}
#pragma endregion

#pragma region Overrides
// Called when the game starts
void UMyBowlPlayerComponent::BeginPlay()
{
	Super::BeginPlay();

	auto _gamemaster = GetGameMaster();
	if (ensure(_gamemaster != nullptr) == false)
	{
		return;
	}

	if (ensure(_gamemaster->OnBallLaunch.IsAlreadyBound(this, &UMyBowlPlayerComponent::StartFollowingBall) == false))
	{
		_gamemaster->OnBallLaunch.AddDynamic(this, &UMyBowlPlayerComponent::StartFollowingBall);
	}
	if (ensure(_gamemaster->OnNudgeBallLeft.IsAlreadyBound(this, &UMyBowlPlayerComponent::NudgeBallLeft) == false))
	{
		_gamemaster->OnNudgeBallLeft.AddDynamic(this, &UMyBowlPlayerComponent::NudgeBallLeft);
	}
	if (ensure(_gamemaster->OnNudgeBallRight.IsAlreadyBound(this, &UMyBowlPlayerComponent::NudgeBallRight) == false))
	{
		_gamemaster->OnNudgeBallRight.AddDynamic(this, &UMyBowlPlayerComponent::NudgeBallRight);
	}
	if (ensure(_gamemaster->BowlNewTurnIsReady.IsAlreadyBound(this, &UMyBowlPlayerComponent::NewTurnIsReady) == false))
	{
		_gamemaster->BowlNewTurnIsReady.AddDynamic(this, &UMyBowlPlayerComponent::NewTurnIsReady);
	}
	if (ensure(_gamemaster->OnWinGame.IsAlreadyBound(this, &UMyBowlPlayerComponent::OnWinGame) == false))
	{
		_gamemaster->OnWinGame.AddDynamic(this, &UMyBowlPlayerComponent::OnWinGame);
	}

	MyStartLocation = GetOwner()->GetActorLocation();
	MyStartRotation = GetOwner()->GetActorRotation();
}


// Called every frame
void UMyBowlPlayerComponent::TickComponent(float DeltaTime, ELevelTick TickType, FActorComponentTickFunction* ThisTickFunction)
{
	Super::TickComponent(DeltaTime, TickType, ThisTickFunction);

	if (bShouldFollowBall && myBall != nullptr)
	{
		auto _myPos = GetOwner()->GetActorLocation();

		if (_myPos.X >= BallFollowLimitDistance)
		{
			bShouldFollowBall = false;
			return;
		}

		auto _ballPos = myBall->GetOwner()->GetActorLocation();
		auto _xTravelPos = _ballPos.X + DefaultBallFollowOffset;
		//PrintString("Ball Pos: " + _ballPos, FLinearColor.Green, printToLog:true);
		GetOwner()->SetActorLocation(
			FVector(_xTravelPos, _myPos.Y, _myPos.Z),
			true, false
		);

	}
}
#pragma endregion

#pragma region Handlers
void UMyBowlPlayerComponent::OnWinGame()
{

}

void UMyBowlPlayerComponent::NewTurnIsReady(EBowlAction _action)
{
	GetOwner()->SetActorLocation(MyStartLocation, false);
	GetOwner()->SetActorRotation(MyStartRotation);
}

void UMyBowlPlayerComponent::StartFollowingBall(FVector launchVelocity, UBowlingBallComponent* bowlingBall)
{
	myBall = bowlingBall;
	if (myBall != nullptr)
	{
		DefaultBallFollowOffset = GetOwner()->GetActorLocation().X - myBall->GetOwner()->GetActorLocation().X;
		bShouldFollowBall = true;
	}
	else
	{
		bShouldFollowBall = false;
	}
}

void UMyBowlPlayerComponent::NudgeBallLeft(float famount)
{
	GetOwner()->SetActorLocation(
		GetOwner()->GetActorLocation() + FVector(0, famount, 0),
		false
	);
}

void UMyBowlPlayerComponent::NudgeBallRight(float famount)
{
	GetOwner()->SetActorLocation(
		GetOwner()->GetActorLocation() + FVector(0, famount, 0),
		false
	);
}
#pragma endregion

#pragma region UFunctions
void UMyBowlPlayerComponent::OnDragStart(FVector2D mousePos)
{
	auto _gamemode = GetBowlGameMode();
	auto _gamemaster = GetGameMaster();
	if(ensure(_gamemode != nullptr) && ensure(_gamemaster != nullptr) &&
		_gamemaster->bCanLaunchBall && _gamemaster->bBowlTurnIsOver == false)
	{
		_gamemode->OnStartDrag(mousePos);
	}
}

void UMyBowlPlayerComponent::OnDragStop(FVector2D mousePos)
{
	auto _gamemode = GetBowlGameMode();
	auto _gamemaster = GetGameMaster();
	if (ensure(_gamemode != nullptr) && ensure(_gamemaster != nullptr) &&
		_gamemaster->bCanLaunchBall && _gamemaster->bBowlTurnIsOver == false)
	{
		_gamemode->OnStopDrag(mousePos);
	}
}
void UMyBowlPlayerComponent::Debug_InstantStrike()
{
	auto _gamemaster = GetGameMaster();
	if(_gamemaster != nullptr)
	{
		_gamemaster->CallDebug_OnSimulateStrike();
	}
}
void UMyBowlPlayerComponent::Debug_Fill18ScoreSlots()
{
	auto _gamemaster = GetGameMaster();
	if (_gamemaster != nullptr)
	{
		_gamemaster->CallDebug_Fill18ScoreSlots();
	}
}
#pragma endregion

File: Managed/OLDREFUSharpBowlProject/BowlingGameComponents/MyBowlPlayerComponent.cs
﻿using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using UnrealEngine.Runtime;
using UnrealEngine.Engine;
using UnrealEngine.Slate;
using UnrealEngine;
using UnrealEngine.GameplayTasks;
using UnrealEngine.SlateCore;
using UnrealEngine.NavigationSystem;
using USharpBowlProject;

namespace OLDREFUSharpBowlProject
{
    [UClass, Blueprintable, BlueprintType]
    public class UMyBowlPlayerComponent : UActorComponent
    {
        #region UProperties
        [UProperty, EditAnywhere, BlueprintReadWrite]
        public string MyString12345 { get; set; }
        [UProperty, EditAnywhere, BlueprintReadWrite, Category("Bowling")]
        public float BallFollowLimitDistance { get; set; }
        [UProperty, BlueprintReadOnly, Category("Bowling")]
        public float DefaultBallFollowOffset { get; set; }
        #endregion

        #region IgnoreProperties
        [UPropertyIgnore]
        protected float PawnStartXPoint { get; set; }
        [UPropertyIgnore]
        public AActor MyOwner
        {
            get
            {
                if (_owner == null)
                    _owner = GetOwner();

                return _owner;
            }
        }
        private AActor _owner = null;
        [UPropertyIgnore]
        protected UBowlGameMasterComponent gamemaster => UBowlGameMasterComponent.GetInstance(MyOwner);
        [UPropertyIgnore]
        protected UBowlGameModeComponent gamemode => UBowlGameModeComponent.GetInstance(MyOwner);
        #endregion

        #region Fields
        protected bool bShouldFollowBall = false;
        private FHitResult myHit;
        private UBowlingBallComponent myBall = null;
        private FVector MyStartLocation;
        private FRotator MyStartRotation;
        #endregion

        #region Testing
        [UFunction, BlueprintCallable]
        public string GetMyMessage()
        {
            return "Hello There From MyPawnClass1234556789";
        }
        #endregion

        #region Overrides
        public override void Initialize(FObjectInitializer initializer)
        {
            //base.Initialize(initializer);
            BallFollowLimitDistance = 3200.0f;
        }

        public override void BeginPlay()
        {
            PawnStartXPoint = MyOwner.GetActorLocation().X;
            gamemaster.OnBallLaunch += StartFollowingBall;
            gamemaster.OnNudgeBallLeft += NudgeBallLeft;
            gamemaster.OnNudgeBallRight += NudgeBallRight;
            gamemaster.BowlNewTurnIsReady += NewTurnIsReady;
            gamemaster.OnWinGame += OnWinGame;

            MyStartLocation = MyOwner.GetActorLocation();
            MyStartRotation = MyOwner.GetActorRotation();
        }

        protected override void ReceiveTick_Implementation(float DeltaSeconds)
        {
            if (bShouldFollowBall && myBall != null)
            {
                var _myPos = MyOwner.GetActorLocation();

                if (_myPos.X >= BallFollowLimitDistance)
                {
                    bShouldFollowBall = false;
                    return;
                }

                var _ballPos = myBall.MyOwner.GetActorLocation();
                var _xTravelPos = _ballPos.X + DefaultBallFollowOffset;
                //PrintString("Ball Pos: " + _ballPos, FLinearColor.Green, printToLog:true);
                MyOwner.SetActorLocation(
                    new FVector(_xTravelPos, _myPos.Y, _myPos.Z),
                    true, out myHit, false
                );

            }
        }

        public override void EndPlay(EEndPlayReason endPlayReason)
        {
            if (gamemaster != null)
            {
                gamemaster.OnBallLaunch -= StartFollowingBall;
                gamemaster.OnNudgeBallLeft -= NudgeBallLeft;
                gamemaster.OnNudgeBallRight -= NudgeBallRight;
                gamemaster.BowlNewTurnIsReady -= NewTurnIsReady;
                gamemaster.OnWinGame -= OnWinGame;
            }
        }
        #endregion

        #region Getters
        [UFunction, BlueprintCallable]
        public FVector GetMyBallPosition()
        {
            if (myBall != null)
            {
                return myBall.MyOwner.GetActorLocation();
            }
            return new FVector(0, 0, 0);
        }

        [UFunction, BlueprintCallable]
        public UBowlGameModeComponent GetBowlGameModeComponent()
        {
            return gamemode;
        }

        [UFunction, BlueprintCallable]
        public UBowlGameMasterComponent GetBowlGameMasterComponent()
        {
            return gamemaster;
        }
        #endregion

        #region Handlers
        void OnWinGame()
        {
            //OnWinGameImplementEvent();
            if (OnWinGameDelegate.IsBound)
            {
                OnWinGameDelegate.Invoke();
            }
        }

        void NewTurnIsReady(EBowlAction _action)
        {
            MyOwner.SetActorLocation(
                MyStartLocation, false, out myHit, false
                );
            MyOwner.SetActorRotation(
                MyStartRotation, false
                );
        } 

        void StartFollowingBall(FVector launchVelocity, UBowlingBallComponent bowlingBall)
        {
            myBall = bowlingBall;
            if (myBall != null)
            {
                DefaultBallFollowOffset = MyOwner.GetActorLocation().X - myBall.MyOwner.GetActorLocation().X;
                bShouldFollowBall = true;
            }
            else
            {
                bShouldFollowBall = false;
            }
        }

        void NudgeBallLeft(float famount)
        {
            FHitResult _hit;
            MyOwner.SetActorLocation(
                MyOwner.GetActorLocation() +
                new FVector(0, famount, 0), false, out _hit, false);
        }

        void NudgeBallRight(float famount)
        {
            FHitResult _hit;
            MyOwner.SetActorLocation(
                MyOwner.GetActorLocation() +
                new FVector(0, famount, 0), false, out _hit, false);
        }
        #endregion

        #region BlueprintImplementableEvents
        ////[UFunction, BlueprintImplementedEvent]
        ////public void OnWinGameImplementEvent()
        ////{

        ////}
        #endregion

        #region TestingDelegates
        [UProperty(PropFlags.BlueprintCallable | PropFlags.BlueprintAssignable), EditAnywhere, BlueprintReadWrite]
        public UBowlGameMasterComponent.FGeneralDelegateHandler OnWinGameDelegate { get; set; }
        #endregion

        #region PublicUFunctions
        [UFunction, BlueprintCallable]
        public void OnDragStart(FVector2D mousePos)
        {
            if (gamemode != null && gamemaster.bCanLaunchBall &&
                gamemaster.bBowlTurnIsOver == false)
            {
                gamemode.OnStartDrag(mousePos);
            }

        }

        [UFunction, BlueprintCallable]
        public void OnDragStop(FVector2D mousePos)
        {
            if (gamemode != null && gamemaster.bCanLaunchBall &&
                gamemaster.bBowlTurnIsOver == false)
            {
                gamemode.OnStopDrag(mousePos);
            }
        }

        [UFunction, BlueprintCallable]
        public void Debug_InstantStrike()
        {
            gamemaster.CallDebug_OnSimulateStrike();
        }

        [UFunction, BlueprintCallable]
        public void Debug_Fill18ScoreSlots()
        {
            gamemaster.CallDebug_Fill18ScoreSlots();
        }
        #endregion
    }
}
